src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.ConstrainedExecution;
+
+namespace System.IO.Enumeration
+{
+    public unsafe abstract partial class FileSystemEnumerator<TResult> : CriticalFinalizerObject, IEnumerator<TResult>
+    {
+        /// <summary>
+        /// Return true if the given file system entry should be included in the results.
+        /// </summary>
+        protected virtual bool ShouldIncludeEntry(ref FileSystemEntry entry) => true;
+
+        /// <summary>
+        /// Return true if the directory entry given should be recursed into.
+        /// </summary>
+        protected virtual bool ShouldRecurseIntoEntry(ref FileSystemEntry entry) => true;
+
+        /// <summary>
+        /// Generate the result type from the current entry;
+        /// </summary>
+        protected abstract TResult TransformEntry(ref FileSystemEntry entry);
+
+        /// <summary>
+        /// Called whenever the end of a directory is reached.
+        /// </summary>
+        /// <param name="directory">The path of the directory that finished.</param>
+        protected virtual void OnDirectoryFinished(ReadOnlySpan<char> directory) { }
+
+        /// <summary>
+        /// Called when a native API returns an error. Return true to continue, or false
+        /// to throw the default exception for the given error.
+        /// </summary>
+        /// <param name="error">The native error code.</param>
+        protected virtual bool ContinueOnError(int error) => false;
+
+        public TResult Current => _current;
+
+        object IEnumerator.Current => Current;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
This is quite a big method. Do you have perf data that proves inlining is worth it?

Reply:
It is only ever used in one place- it was an optimization for code sharing between the various platforms to put it in a method. @jkotas- is there a downside to doing this when you know it is logically "embedded"?

Reply:
If you keep this it could use a comment explaining

Reply:
- Merging two logically embedded methods into one is not always a win.
- It bit rots as the code is changing. It makes the code hard to maintain because.

AggressiveInlining should be only used when you have measured that it improves performance for something you care about. It does not seem to be the case here.

Otherwise, leave it to the compiler to decide.

Reply:
Ok then, pulling it.